metadata/reader_test.go

```diff
@@ -33,47 +33,262 @@
+				testutil.Dir("dir1/"),
+				testutil.File("dir1/../dir1///////////////////file1", ""),
+				testutil.Chardev("dir1/cdev", 10, 11),
+				testutil.Blockdev("dir1/bdev", 100, 101),
+				testutil.Fifo("dir1/fifo"),
+			},
+			want: []check{
+				numOfNodes(6), // root dir + 1 file + 1 dir + 1 cdev + 1 bdev + 1 fifo
+				hasFile("dir1/file1", 0),
+				hasChardev("dir1/cdev", 10, 11),
+				hasBlockdev("dir1/bdev", 100, 101),
+				hasFifo("dir1/fifo"),
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		for _, prefix := range allowedPrefix {
+			prefix := prefix
+			for srcCompresionName, srcCompression := range srcCompressions {
+				srcCompresionName, srcCompression := srcCompresionName, srcCompression
+				t.Run(tt.name+"-"+srcCompresionName, func(t *testing.T) {
+					opts := []testutil.BuildTarOption{
+						testutil.WithPrefix(prefix),
+					}
+
+					ztoc, sr, err := ztoc.BuildZtocReader(t, tt.in, srcCompression, 64, opts...)
+					if err != nil {
+						t.Fatalf("failed to build ztoc: %v", err)
+					}
+					telemetry, checkCalled := newCalledTelemetry()
+
+					// create a metadata reader
+					r, err := newTestableReader(sr, ztoc.TOC, WithTelemetry(telemetry))
+					if err != nil {
+						t.Fatalf("failed to create new reader: %v", err)
+					}
+					defer r.Close()
+					t.Logf("vvvvv Node tree vvvvv")
+					t.Logf("[%d] ROOT", r.RootID())
+					dumpNodes(t, r, r.RootID(), 1)
+					t.Logf("^^^^^^^^^^^^^^^^^^^^^")
+					for _, want := range tt.want {
+						want(t, r)
+					}
+					if err := checkCalled(); err != nil {
+						t.Errorf("telemetry failure: %v", err)
+					}
+				})
+			}
+		}
+	}
 }
 
-func newTestableReader(sr *io.SectionReader, toc ztoc.TOC, opts ...Option) (testableReader, error) {
-	f, err := os.CreateTemp("", "readertestdb")
-	if err != nil {
-		return nil, err
+func BenchmarkMetadataReader(b *testing.B) {
+	testCases := []struct {
+		name    string
+		entries int
+	}{
+		{
+			name:    "Create metadata.Reader with 1,000 TOC entries",
+			entries: 1000,
+		},
+		{
+			name:    "Create metadata.Reader with 10,000 TOC entries",
+			entries: 10_000,
+		},
+		{
+			name:    "Create metadata.Reader with 50,000 TOC entries",
+			entries: 50_000,
+		},
+		{
+			name:    "Create metadata.Reader with 100,000 TOC entries",
+			entries: 100_000,
+		},
 	}
-	defer os.Remove(f.Name())
-	db, err := bolt.Open(f.Name(), 0600, nil)
+	cwdPath, err := os.Getwd()
 	if err != nil {
-		return nil, err
+		b.Fatal(err)
 	}
-	r, err := NewReader(db, sr, toc, opts...)
-	if err != nil {
-		return nil, err
+	for _, tc := range testCases {
+		tempDB, clean, err := newTempDB(cwdPath)
+		defer clean()
+		if err != nil {
+			b.Fatalf("failed to initialize temp db: %v", err)
+		}
+		toc, err := generateTOC(tc.entries)
+		if err != nil {
+			b.Fatalf("failed to generate TOC: %v", err)
+		}
+		b.ResetTimer()
+		b.Run(tc.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				if _, err := NewReader(tempDB, nil, toc); err != nil {
```

Comment:
Does this rely on the fact that `NewReader` will write new entries to the `tempDB` even if this TOC has already been used?

Reply:
It does, yeah. I suppose I could initialize a new DB per-run with `b.Start` and `b.Stop`.